### PR TITLE
Move forms and instances data code to module

### PR DESCRIPTION
**Pull request by seadowg**
A lot of this PR ends up being just moves and changes to imports which can be a little hard to follow, so I'd advise going commit by commit. I'll also lay out the key changes here:

* Introduced providers for the forms and instances repositories (lets them not have zero arg constructors and will be needed for projects)
* Move Form and Instances pure Java (non-Android) data code to a new `forms` module
* Move test helpers for the above (`FormsUtils` and `FormsRepositoryTest` for example) to a new `formstest` module
* Create a new `shared` module for common non-Android code (and renamed the old `shared` to `androidshared`)
* Reorganized the `databases` package so forms and instances have the same structure

There are some things I would have liked to do but didn't get time to (I gave myself a one-day timebox for this). Let me know if you think any of these should go in for completeness: 

1. Move the `database` package out to a new module
1. Move the forms and instances content provider implementations to their own Android library module

It's worth noting that these are in order as each change would depend on the last. Another possible positive change we could make now is to move `openrosa` (our open rosa based implementation of `FormSource`) out to a new module.

#### What has been done to verify that this works as intended?

Ran tests.

#### Why is this the best possible solution? Were any other approaches considered?

I wanted to specifically move the abstractions and data classes around forms and instances out to give us scope to move other parts of the code out to their own modules (form management, content providers etc). Most of the other changes are simply facilitating that.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

I think this is pretty safe as it's mostly just moving things around. Shouldn't need QA if all checks are passing.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by seadowg** [comment]
I've added a `README.md` to `formstest` to try and explain what it's for.

File: collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormColumns.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        values.put(FORM_MEDIA_PATH, storagePathProvider.getRelativeFormPath(form.formMediaPath))
+        values.put(LANGUAGE, form.language)
+        values.put(AUTO_SEND, form.autoSend)
+        values.put(AUTO_DELETE, form.autoDelete)
+        values.put(GEOMETRY_XPATH, form.geometryXpath)
+        return values
+    }
+
+    @JvmStatic
+    fun getFormFromValues(values: ContentValues, storagePathProvider: StoragePathProvider): Form? {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormColumns.kt]: Is this class really a good place for method like this one? This was a class for constants rather.

File: collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormsRepository.java [hunk 2]
```diff
@@ -38,10 +38,10 @@
 public class DatabaseFormsRepository implements FormsRepository {
 
     private final StoragePathProvider storagePathProvider;
-    private final Clock clock;
+    private final Supplier<Long> clock;
     private final FormsDatabaseProvider formsDatabaseProvider;
 
-    public DatabaseFormsRepository(Clock clock, StoragePathProvider storagePathProvider, FormsDatabaseProvider formsDatabaseProvider) {
+    public DatabaseFormsRepository(Supplier<Long> clock, StoragePathProvider storagePathProvider, FormsDatabaseProvider formsDatabaseProvider) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormsRepository.java:44]: This change will work assuming that we create a repository object every time we want to use it. OK maybe now this is the case but what if at some point someone reuses an already created object? Isn't it a bit dangerous?
What's the advantage of such a change?

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]
We have more and more modules. It would be good to have a graphical representation of them so that we can just take a look and understand the dependencies between them.

File: collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormColumns.kt (hunk 1 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormColumns.kt]: It seemed to make sense to me as it you often end up needing both the columns and the conversion between `ContentValues`/`Cursor` in the same place. I do see what you mean though. You think another "helpers" or "utils" is in order?

File: collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormsRepository.java (hunk 2 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormsRepository.java:44]: The advantage for me is using a standard library abstraction rather than a custom one - we can probably get rid of `Clock` tbh. What is the worry around the object being reused?

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormsRepository.java:44]: The worry is that if you pass long value when you create an object it's a constant and if you reuse the object later that value is not you current timestamp anymore so wrong timestamps might be saved to the database?

File: collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormColumns.kt (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormColumns.kt]: Yeah to me it seems odd and like something I wouldn't do so it cough my attention.

File: collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormsRepository.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormsRepository.java:44]: Ahh my mistake that's not the way it works probably now you are wondering what I was talking about sorry it won't be outdated.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormsRepository.java:44]: Ah I see what you were worried about! Right yeah as you say, we still get a different value each time, so we're ok. Definitely a good one to look out for though.

File: collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormColumns.kt (hunk 1 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/database/forms/DatabaseFormColumns.kt]: Yeah happy to move a helper out if it feels wrong! I'll do that.

File: .circleci/config.yml [hunk 3]
```diff
@@ -17,19 +17,23 @@ jobs:
     steps:
       - checkout
 
+      - run:
```
Comment (seadowg) [inline .circleci/config.yml:20]: Added some proper caching based on dependencies changing to help speed up CI. I was running into problems with timeouts so thought it would be food to fix while I'm here.

**Review by grzesiek2010 — APPROVED** [review]
